src/udev/udevd.c

```diff
@@ -1544,59 +1554,69 @@ static int on_sighup(sd_event_source *s, const struct signalfd_siginfo *si, void
+        sd_device *dev;
         int r;
 
-        assert(manager);
+        assert(si);
 
-        for (;;) {
-                pid_t pid;
-                int status;
-                Worker *worker;
+        worker = hashmap_get(manager->workers, PID_TO_PTR(si->si_pid));
+        if (!worker) {
+                log_warning("Worker ["PID_FMT"] is unknown, ignoring.", si->si_pid);
+                return 0;
+        }
 
-                pid = waitpid(-1, &status, WNOHANG);
-                if (pid <= 0)
-                        break;
+        dev = worker->event ? ASSERT_PTR(worker->event->dev) : NULL;
 
-                worker = hashmap_get(manager->workers, PID_TO_PTR(pid));
-                if (!worker) {
-                        log_warning("Worker ["PID_FMT"] is unknown, ignoring", pid);
-                        continue;
-                }
+        switch (si->si_code) {
+        case CLD_EXITED:
+                if (si->si_status == 0)
+                        log_device_debug(dev, "Worker ["PID_FMT"] exited.", si->si_pid);
+                else
+                        log_device_warning(dev, "Worker ["PID_FMT"] exited with return code %i.",
+                                           si->si_pid, si->si_status);
+                result = EVENT_RESULT_EXIT_STATUS_BASE + si->si_status;
+                break;
 
-                if (WIFEXITED(status)) {
-                        if (WEXITSTATUS(status) == 0)
-                                log_debug("Worker ["PID_FMT"] exited", pid);
-                        else
-                                log_warning("Worker ["PID_FMT"] exited with return code %i", pid, WEXITSTATUS(status));
-                } else if (WIFSIGNALED(status))
-                        log_warning("Worker ["PID_FMT"] terminated by signal %i (%s)", pid, WTERMSIG(status), signal_to_string(WTERMSIG(status)));
-                else if (WIFSTOPPED(status)) {
-                        log_info("Worker ["PID_FMT"] stopped", pid);
-                        continue;
-                } else if (WIFCONTINUED(status)) {
-                        log_info("Worker ["PID_FMT"] continued", pid);
-                        continue;
-                } else
-                        log_warning("Worker ["PID_FMT"] exit with status 0x%04x", pid, status);
+        case CLD_KILLED:
+        case CLD_DUMPED:
+                log_device_warning(dev, "Worker ["PID_FMT"] terminated by signal %i (%s).",
+                                   si->si_pid, si->si_status, signal_to_string(si->si_status));
+                result = EVENT_RESULT_SIGNAL_BASE + si->si_status;
+                break;
+
+        case CLD_TRAPPED:
+                log_device_info(dev, "Worker ["PID_FMT"] trapped.", si->si_pid);
+                return 0;
 
-                if ((!WIFEXITED(status) || WEXITSTATUS(status) != 0) && worker->event) {
-                        log_device_error(worker->event->dev, "Worker ["PID_FMT"] failed", pid);
+        case CLD_STOPPED:
```

Comment:
these two cannot happen given that you passed WEXITED to sd_event_add_child() and not `WEXITED|WCONTINUED|WSTOPPED`.

hence, maybe just add `assert_nort_reached()` for these two

Reply:
How about CLD_TRAPPED?

Reply:
Thanks. 

BTW, where can I get the official information about that? The man page wait(2) does not explicitly explain the relation CLD_XXX code vs WXXX flag, especially for trapped case.

Reply:
yeah, it's not explicitly documented. But you can derive it from the description of WIFSTOPPED() in a way...